pkg/k8s/informers/informers.go

```diff
@@ -0,0 +1,167 @@
+)
+
+type InformerCollectionOpt func(*InformerCollection)
+
+func NewInformerCollection(meshName string, stop <-chan struct{}, kubeClient kubernetes.Interface, smiTrafficSplitClient smiTrafficSplitClient.Interface,
+	smiTrafficSpecClient smiTrafficSpecClient.Interface, smiAccessClient smiTrafficAccessClient.Interface, configClient configClientset.Interface,
+	policyClient policyClientset.Interface, opts ...InformerCollectionOpt) (*InformerCollection, error) {
+
+	ic := &InformerCollection{
+		meshName:              meshName,
+		kubeClient:            kubeClient,
+		smiTrafficSplitClient: smiTrafficSplitClient,
+		smiTrafficSpecClient:  smiTrafficSpecClient,
+		smiAccessClient:       smiAccessClient,
+		configClient:          configClient,
+		policyClient:          policyClient,
+		informers:             make(informerCollection),
+	}
+
+	for _, opt := range opts {
+		opt(ic)
+	}
+
+	// Initialize informers
+	informerInitHandlerMap := map[InformerKey]informerInit{
+		// Kubernetes
+		InformerKeyNamespace:      ic.initNamespaceMonitor,
+		InformerKeyService:        ic.initServicesMonitor,
+		InformerKeyServiceAccount: ic.initServiceAccountsMonitor,
+		InformerKeyPod:            ic.initPodMonitor,
+		InformerKeyEndpoints:      ic.initEndpointMonitor,
+
+		// SMI
+		InformerKeyTrafficSplit:   ic.initTrafficSplitMonitor,
+		InformerKeyTrafficTarget:  ic.initTrafficTargetMonitor,
+		InformerKeyHTTPRouteGroup: ic.initHTTPRouteGroupMonitor,
+		InformerKeyTCPRoute:       ic.initTCPRouteMonitor,
+
+		// Config
+		InformerKeyMeshConfig:          ic.initMeshConfigMonitor,
+		InformerKeyMeshRootCertificate: ic.initMeshRootCertificateMonitor,
+
+		// Policy
+		InformerKeyEgress:                 ic.initEgressMonitor,
+		InformerKeyIngressBackend:         ic.initIngressBackendMonitor,
+		InformerKeyUpstreamTrafficSetting: ic.initUpstreamTrafficSettingMonitor,
+		InformerKeyRetry:                  ic.initRetryMonitor,
+	}
+
+	if len(ic.selectedInformers) == 0 {
+		// Initialize all informers
+		ic.selectedInformers = []InformerKey{
+			InformerKeyNamespace,
+			InformerKeyService,
+			InformerKeyPod,
+			InformerKeyEndpoints,
+			InformerKeyServiceAccount,
+			InformerKeyTrafficSplit,
+			InformerKeyTrafficTarget,
+			InformerKeyHTTPRouteGroup,
+			InformerKeyTCPRoute,
+			InformerKeyMeshConfig,
+			InformerKeyMeshRootCertificate,
+			InformerKeyEgress,
+			InformerKeyIngressBackend,
+			InformerKeyUpstreamTrafficSetting,
+			InformerKeyRetry,
+		}
+	}
+
+	for _, key := range ic.selectedInformers {
+		informerInitHandlerMap[key]()
+	}
+
+	if err := ic.run(stop); err != nil {
+		log.Error().Err(err).Msg("Could not start informer collection")
+		return nil, err
+	}
+
+	return ic, nil
+}
+
+// WithCustomStore sets the shared store for the informerCollection to reference.
+// This store will be used instead of each informer's store. This should typically
+// only be used in tests
+func WithCustomStores(stores map[InformerKey]cache.Store) InformerCollectionOpt {
```

Comment:
```// WithCustomStores provides the InformerCollection an set of `cache.Store`s indexed
// by InformerKey. This functionality was added for the express purpose of testing
// flexibility since the alternative often leads to flaky tests and race conditions
// between the time an object is added to a fake clientset and when that object
// is actually added to the informer `cache.Store`.```